state-chain/chains/src/ccm_checker.rs

```diff
@@ -0,0 +1,332 @@
+use crate::{
+	sol::{
+		api::SolanaEnvironment, SolAsset, SolCcmAccounts, CCM_BYTES_PER_ACCOUNT, MAX_CCM_BYTES_SOL,
+		MAX_CCM_BYTES_USDC,
+	},
+	CcmChannelMetadata, CcmValidityCheck, CcmValidityError,
+};
+use cf_primitives::{Asset, ForeignChain};
+use codec::Decode;
+use core::marker::PhantomData;
+
+pub struct CcmValidityChecker<Environment> {
+	_phantom: PhantomData<Environment>,
+}
+
+impl<Environment: SolanaEnvironment> CcmValidityCheck for CcmValidityChecker<Environment> {
+	fn is_valid(ccm: &CcmChannelMetadata, egress_asset: Asset) -> Result<(), CcmValidityError> {
+		if ForeignChain::from(egress_asset) == ForeignChain::Solana {
+			// Check if the cf_parameter can be decoded
+			let ccm_accounts = SolCcmAccounts::decode(&mut &ccm.cf_parameters.clone()[..])
+				.map_err(|_| CcmValidityError::CannotDecodeCfParameters)?;
+			let asset: SolAsset = egress_asset
+				.try_into()
+				.expect("Only Solana chain's asset will be checked. This conversion must succeed.");
+
+			// Length of CCM = length of message + total no. remaining_accounts * constant;
+			let ccm_length =
+				ccm.message.len() + ccm_accounts.remaining_accounts.len() * CCM_BYTES_PER_ACCOUNT;
+			if ccm_length >
+				match asset {
+					SolAsset::Sol => MAX_CCM_BYTES_SOL,
+					SolAsset::SolUsdc => MAX_CCM_BYTES_USDC,
+				} {
+				return Err(CcmValidityError::CcmIsTooLong)
+			}
+
+			// Check if the parameter accounts are valid
+			if let Ok(api_env) = Environment::api_environment() {
+				let token_pda = api_env.token_vault_pda_account.into();
+				if ccm_accounts.cf_receiver.pubkey == token_pda ||
+					ccm_accounts.remaining_accounts.iter().any(|acc| acc.pubkey == token_pda)
+				{
+					return Err(CcmValidityError::CfParametersContainsInvalidAccounts)
+				}
+			};
+			if let Ok(agg_key) = Environment::current_agg_key() {
+				let agg_key = agg_key.into();
+				if ccm_accounts.cf_receiver.pubkey == agg_key ||
+					ccm_accounts.remaining_accounts.iter().any(|acc| acc.pubkey == agg_key)
+				{
+					return Err(CcmValidityError::CfParametersContainsInvalidAccounts)
+				}
+			};
+
+			Ok(())
+		} else {
+			Ok(())
+		}
+	}
+}
```

Comment:
I dont think we need to make the ```CcmValidityChecker``` generic over the Environment. We inject it in the swapping pallet which is not instantiable and so doesnt need to be bound to a specific chain. If we need a checker for any other chain we have to add the logic of the checker in the same type (because swapping pallet is not instantiable). So basically, the same ```is_valid``` function will have the logic for the checker for other chains as well with a match statement.